### PR TITLE
Fixed problem with Pretty printing adding semicolon

### PR DESCRIPTION
**Pull request by klangner**
I decided to add new tests based on HUnit.
Without tests I'm afraid we will broke this functionality again in the future.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,6 @@
 /dist/
 *.out
+.cabal-sandbox/
+.idea/
```
Comment (mpickering) [inline .gitignore]: What is this entry for?

File: haskell-src-exts.cabal [hunk 2]
```diff
@@ -88,6 +88,7 @@ Test-Suite test
                        tasty >= 0.3,
                        tasty-smallcheck,
                        tasty-golden >= 2.2.2,
+                       tasty-hunit,
```
Comment (mpickering) [inline haskell-src-exts.cabal]: Why did you need to add this dependency? Usually we must have quite a good reason to add a dependency but this might be ok as it is only for the tests. I would still like to understand why it is necessary.

File: src/Language/Haskell/Exts/Pretty.hs [hunk 3]
```diff
@@ -490,8 +490,10 @@ instance  Pretty (Decl l) where
 
         pretty (FunBind _ matches) = do
                 e <- fmap layout getPPEnv
-                case e of PPOffsideRule -> foldr ($$$) empty (map pretty matches)
-                          _ -> foldr (\x y -> x <> semi <> y) empty (map pretty matches)
+                case e of
+                    PPOffsideRule -> foldr ($$$) empty (map pretty matches)
+                    PPInLine -> foldr ($$$) empty (map pretty matches)
```
Comment (mpickering) [inline src/Language/Haskell/Exts/Pretty.hs]: I don't think this fixes the problem properly. The `PPInLine` option is meant to print

```
foo a = 1
foo b = 2
```
as

```
foo a = 1; foo b = 2
```

which is not what happens with this patch.

The correct fix is to change line 494 from

```
_ -> foldr (\x y -> x <> semi <> y) empty (map pretty matches)
```
to
```
_ -> hsep $ punctuate semi (map pretty matches)
```

which will deal with the `n=0` case correctly.

File: tests/Invariants.hs [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+import Test.Tasty (TestTree, testGroup)
+import Test.Tasty.HUnit
+import Language.Haskell.Exts
+
+
+invariantTests :: TestTree
+invariantTests = testGroup "Unit tests"
+    [ testCase "Pretty printing with layout = PPInLine" $ do
+        let decl = "h x = f (filter ('a' ==) x)"
```
Comment (mpickering) [inline tests/Invariants.hs]: Add a test here for the example I gave in the other comment.

File: .gitignore (hunk 1 above)
Comment (klangner) [inline .gitignore]: The first one is for cabal sandbox temporary files. 
The second is for Intellij IDE.
I added this so this file won't be committed to the repository.
I hope it is ok since those files should never be added to the git repo?

BTW I though about using stack instead of cabal sandbox, but that's probably should be done in separate task.

File: haskell-src-exts.cabal (hunk 2 above)
Comment (klangner) [inline haskell-src-exts.cabal]: I agree that it is not good idea to add dependency.

I did it because AFAIK there is no library for unit tests in this project. I though about using property based testing, but after your suggestion that that could not be good idea, decided to write unit tests instead. 
But probably I should discuss this before preparing this PR.

File: src/Language/Haskell/Exts/Pretty.hs (hunk 3 above)
Comment (klangner) [inline src/Language/Haskell/Exts/Pretty.hs]: Yes you right.

File: tests/Invariants.hs (hunk 4 above)
Comment (klangner) [inline tests/Invariants.hs]: Ok. 
Just let me know if the changes to .gitignore and .cabal files are ok, or what to do with them and I will prepare new PR.

**Comment by mpickering** [comment]
I applied the fix I suggested in an earlier comment.
